### PR TITLE
Refactor AppData data collection and add JVM serialization test

### PR DESCRIPTION
**Pull request by fractalwrench**
## Changeset

- Rename `getAppMetadata()` for consistency with `Metadata` capitalization
- Use `config.versionCode` directly within `AppData` to simplify implementation
- Move serialization tests from instrumentation sourceset to a JVM test, consolidating in one class
- Enabled mockmaker-inline, which is required for Mockito to mock final classes

### CODE REVIEW

File: bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java [hunk 1]
```diff
@@ -63,20 +63,4 @@ public void testAccessors() {
...
-    public void testJsonSerialisation() throws JSONException {
-        appData.put("buildUUID", "fa54de");
-        JSONObject appDataJson = mapToJson(appData);
-
-        assertEquals(1, appDataJson.getInt("versionCode"));
-        assertEquals("1.0", appDataJson.get("version"));
-        assertEquals("development", appDataJson.get("releaseStage"));
-        assertEquals("android", appDataJson.get("type"));
-        assertEquals("com.bugsnag.android.core.test", appDataJson.get("id"));
-        assertNotNull(appDataJson.get("buildUUID"));
```
Comment (simonbowring) [inline bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java]: Do we need to check the serialization of `buildUUIID`, `duration`, `durationInForeground` and `inForeground` in the new tests like this test was?

Comment (tomlongridge) [reply bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java]: It feels like we have a lot of different test classes - AppDataSummaryTest, AppDataTest, AppDataOverrideTest and AppDataSerialisationTest - but they don't seem to cover all the fields between them.

Can these be rationalised and make sure that we include each of the fields of App?

Comment (fractalwrench) [reply bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java]: I have removed the various different test classes in favour of `AppDataSerializationTest`, which now covers the serialization of each value defined within AppData.

Comment (fractalwrench) [reply bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java]: This is also an issue with tests for `DeviceData` - I'll address that in a separate PR.

File: bugsnag-android-core/src/main/java/com/bugsnag/android/AppData.java [hunk 2]
```diff
@@ -191,11 +186,8 @@ private long getMemoryUsage() {
      * Check if the device is currently running low on memory.
      */
     @Nullable
-    private Boolean isLowMemory() {
+    private Boolean isLowMemory(ActivityManager activityManager) {
```
Comment (tomlongridge) [inline bugsnag-android-core/src/main/java/com/bugsnag/android/AppData.java]: This doesn't need to be passed in as a parameter - it can use the member variable.

File: bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            // construct AppData object
+            val appData = AppData(context, pm, convert(config), sessionTracker, am)
+            appData.setBinaryArch("x86")
+
+            // serializes the 3 different maps that AppData can generate:
+            // 1. summary (used in session payloads)
+            // 2. regular (used in event payloads)
+            // 3. metadata (used in event payloads)
+            val metadata = appData.appDataMetadata
+            metadata.remove("memoryUsage")
```
Comment (tomlongridge) [inline bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt]: Can this not be tested using a mock?

File: bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java (hunk 1 above)
Comment (tomlongridge) [reply bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java]: Should we not also have basic unit tests for the accessors (separate from serialisation)? They would be a unit test, not an instrumentation test, but would check the getters return an appropriate value.

File: bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt (hunk 3 above)
Comment (fractalwrench) [reply bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt]: I don't believe this is possible with the current implementation as this calculation uses `Runtime`, which is not mockable as it should only be instantiated once per application.

Comment (tomlongridge) [reply bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt]: This makes me more confident that we should have a more basic unit test for the getters in AppData - then we can at least assert that it's non-null?

https://github.com/bugsnag/bugsnag-android/pull/661#discussion_r351794414

File: bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java (hunk 1 above)
Comment (fractalwrench) [reply bugsnag-android-core/src/androidTest/java/com/bugsnag/android/AppDataTest.java]: `AppData#getAppDataSummary()` returns a `Map`, so I'm not sure that testing the accessors would be necessary as we're not defining any of our own. Unless I've got the wrong end of the stick here?

File: bugsnag-android-core/src/main/java/com/bugsnag/android/AppData.java (hunk 2 above)
Comment (fractalwrench) [reply bugsnag-android-core/src/main/java/com/bugsnag/android/AppData.java]: Updated to use the field.

File: bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt (hunk 3 above)
Comment (tomlongridge) [reply bugsnag-android-core/src/test/java/com/bugsnag/android/AppDataSerializationTest.kt]: I'll raise this separately to address unit test structure for this and Device.
